### PR TITLE
games-fps/gzdoom: GZDoom requires FluidSynth version 1

### PR DESCRIPTION
**Pull request by vilhelmgray**
Closes: https://bugs.gentoo.org/687716
Package-Manager: Portage-2.3.67, Repoman-2.3.14
Signed-off-by: William Breathitt Gray <vilhelm.gray@gmail.com>

### CODE REVIEW

**Comment by a17r** [comment]
There is no reason to do this in a PR separate from https://github.com/gentoo/gentoo/pull/12227 and cause twice the CI load.

**Comment by a17r** [comment]
Here's what openSUSE does wrt fluidsynth-2: https://build.opensuse.org/package/view_file/games/gzdoom/fl2.patch?expand=1

**Comment by vilhelmgray** [comment]
> Here's what openSUSE does wrt fluidsynth-2: https://build.opensuse.org/package/view_file/games/gzdoom/fl2.patch?expand=1

Despite its commit message, I don't think that patch is actually fixing GZDoom to support FluidSynth2 -- GZDoom still tries to [load `libfluidsynth.so.1` as defined by line 66](https://github.com/coelckers/gzdoom/blob/master/src/sound/mididevices/music_fluidsynth_mididevice.cpp#L66) and the rest of file has been coded with the FluidSynth 1 API in mind.

Regardless, I've ping upstream for comment: https://github.com/coelckers/gzdoom/pull/865#issuecomment-500356141

**Comment by vilhelmgray** [comment]
>     * if `4.1.3` is going to be the new version containing the fluidsynth fix, is anyone actually going to use `4.1.2-r1`?

That's a fair point -- I think most users will want to upgrade to the latest release anyway. Should I drop the existing `4.1.2` ebuild as well, leaving just the new `4.1.3` in the repository?

**Comment by a17r** [comment]
> > ```
> > * if `4.1.3` is going to be the new version containing the fluidsynth fix, is anyone actually going to use `4.1.2-r1`?
> > ```
> 
> That's a fair point -- I think most users will want to upgrade to the latest release anyway. Should I drop the existing `4.1.2` ebuild as well, leaving just the new `4.1.3` in the repository?

If we manage to support fluidsynth-2 (maybe through a combined opensuse+your patch) then it would make sense to keep 4.1.2 for a bit longer, if we just fix the fluidsynth dependency to <2, then I'm for dropping 4.1.2 in a second commit here.

**Comment by vilhelmgray** [comment]
> > > ```
> > > * if `4.1.3` is going to be the new version containing the fluidsynth fix, is anyone actually going to use `4.1.2-r1`?
> > > ```
> > 
> > 
> > That's a fair point -- I think most users will want to upgrade to the latest release anyway. Should I drop the existing `4.1.2` ebuild as well, leaving just the new `4.1.3` in the repository?
> 
> If we manage to support fluidsynth-2 (maybe through a combined opensuse+your patch) then it would make sense to keep 4.1.2 for a bit longer, if we just fix the fluidsynth dependency to <2, then I'm for dropping 4.1.2 in a second commit here.

I'm not convinced yet that the openSUSE patch actually resolves this issue, but I'll reach out to the package maintainer to understand better what's happening there.

I'm going to add a second commit here to drop `4.1.2` anyway since it was only released last month and likely has few users.

**Comment by a17r** [comment]
> I'm not convinced yet that the openSUSE patch actually resolves this issue, but I'll reach out to the package maintainer to understand better what's happening there.

Thanks for taking the time. Getting that done would be much appreciated as `<` version limitations are highly irritating for the users and it would be the only package blocking us from eventual `fluidsynth-1*` cleanup.

**Comment by vilhelmgray** [comment]
> > I'm not convinced yet that the openSUSE patch actually resolves this issue, but I'll reach out to the package maintainer to understand better what's happening there.
> 
> Thanks for taking the time. Getting that done would be much appreciated as `<` version limitations are highly irritating for the users and it would be the only package blocking us from eventual `fluidsynth-1*` cleanup.

I spoke with Jan Engelhardt <jengelh@inai.de> over email, and discovered the openSUSE package is simply disabling the dlopen code and linking with FluidSynth 2.

See discussion about GZDoom handling shared objects: https://github.com/coelckers/gzdoom/pull/528#issuecomment-430599320

According to Jan, there were no API changes between FluidSynth 1 and FluidSynth 2 that affect GZDoom code; so that's how openSUSE is able to run like that, despite the upstream GZDoom team coding to just FluidSynth 1.

However, we have to decide now how to move forward with the Gentoo package:
* we can follow the openSUSE package and link to FluidSynth 2
* or stay on FluidSynth 1 with the upstream GZDoom project

The GZDoom team seems to prefer maintaining their own versions of external libraries within their code repository, so they are unlikely to move forward to FluidSynth 2. That means we'll likely have to maintain downstream patches to support FluidSynth 2 as the API inevitable diverges in the long term future.

I suppose there's not much choice if FluidSynth 1 will be dropped from Gentoo in the near future. If you think it's the right path to take, I'll incorporate the openSUSE patch and let the ebuild pick up FluidSynth 2 as the dependency.

**Comment by a17r** [comment]
> The GZDoom team seems to prefer maintaining their own versions of external libraries within their code repository, so they are unlikely to move forward to FluidSynth 2. That means we'll likely have to maintain downstream patches to support FluidSynth 2 as the API inevitable diverges in the long term future.

I doubt maintaining copies of all their dependencies is *really* what they want, we should lobby them to properly move to FS-2 if it is no big effort.

> I suppose there's not much choice if FluidSynth 1 will be dropped from Gentoo in the near future. If you think it's the right path to take, I'll incorporate the openSUSE patch and let the ebuild pick up FluidSynth 2 as the dependency.

The other option would be to drop GZDoom eventually, if upstream is unwilling to get it right.

**Comment by vilhelmgray** [comment]
I've submitted the openSUSE patch to upstream (hopefully they'll merge it): https://github.com/coelckers/gzdoom/pull/865#issuecomment-500356141

I've included the patch in my latest force-push, and also changed the ebuild dependency back so that fluidsynth-2 can be used.

File: games-fps/gzdoom/gzdoom-4.1.3.ebuild [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+	media-libs/libsdl2[opengl]
+	sys-libs/zlib
+	virtual/jpeg:0
+	gtk? (
+		gtk2? ( x11-libs/gtk+:2 )
+		!gtk2? ( x11-libs/gtk+:3 )
+	)"
+RDEPEND="
+	${DEPEND}
+	fluidsynth? ( media-sound/fluidsynth )
```
Comment (a17r) [inline games-fps/gzdoom/gzdoom-4.1.3.ebuild]: missing slot operator, will fix on merge.

**Pull request closed without merging**